File: command/command.kt
// Command Interface
interface Command {
    fun execute()
    fun undo()
}

// Concrete Command for Adding
class AddCommand(private val list: MutableList<Int>, private val value: Int) : Command {
    override fun execute() {
        list.add(value)
    }

    override fun undo() {
        list.removeAt(list.size - 1)
    }
}

// Invoker
class Calculator {
    private val commandHistory: MutableList<Command> = mutableListOf()
    private var currentCommandIndex: Int = -1

    fun executeCommand(command: Command) {
        commandHistory.add(command)
        command.execute()
        currentCommandIndex++
    }

    fun undo() {
        if (currentCommandIndex >= 0) {
            commandHistory[currentCommandIndex].undo()
            currentCommandIndex--
        }
    }

    fun redo() {
        if (currentCommandIndex < commandHistory.size - 1) {
            currentCommandIndex++
            commandHistory[currentCommandIndex].execute()
        }
    }
}

// Client Code
fun main() {
    val list = mutableListOf(1, 2, 3)
    val calculator = Calculator()

    val addFive = AddCommand(list, 5)
    val addTen = AddCommand(list, 10)

    println("Initial list: $list")  // Output: Initial list: [1, 2, 3]

    calculator.executeCommand(addFive)
    println("After adding 5: $list")  // Output: After adding 5: [1, 2, 3, 5]

    calculator.executeCommand(addTen)
    println("After adding 10: $list")  // Output: After adding 10: [1, 2, 3, 5, 10]

    calculator.undo()
    println("After undo: $list")  // Output: After undo: [1, 2, 3, 5]

    calculator.redo()
    println("After redo: $list")  // Output: After redo: [1, 2, 3, 5, 10]
}

File: command/command.py
class Command:
    def execute(self):
        pass

    def undo(self):
        pass

# Concrete Command for Adding
class AddCommand(Command):
    def __init__(self, lst, value):
        self.lst = lst
        self.value = value

    def execute(self):
        self.lst.append(self.value)

    def undo(self):
        self.lst.pop()

# Invoker
class Calculator:
    def __init__(self):
        self.command_history = []
        self.current_command_index = -1

    def execute_command(self, command):
        self.command_history.append(command)
        command.execute()
        self.current_command_index += 1

    def undo(self):
        if self.current_command_index >= 0:
            self.command_history[self.current_command_index].undo()
            self.current_command_index -= 1

    def redo(self):
        if self.current_command_index < len(self.command_history) - 1:
            self.current_command_index += 1
            self.command_history[self.current_command_index].execute()

# Client code
if __name__ == "__main__":
    lst = [1, 2, 3]
    calculator = Calculator()

    add_five = AddCommand(lst, 5)
    add_ten = AddCommand(lst, 10)

    print(f"Initial list: {lst}")

    calculator.execute_command(add_five)
    print(f"After adding 5: {lst}")

    calculator.execute_command(add_ten)
    print(f"After adding 10: {lst}")

    calculator.undo()
    print(f"After undo: {lst}")

    calculator.redo()
    print(f"After redo: {lst}")
